documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,438 @@
+     * @param lastWrittenPosition The position the writer was at when it noted the time.
+     */
+    void noteTimestampFromWriter(String writer, Stream stream, long timestamp, Position lastWrittenPosition);
+
+    /**
+     * Notifies the controller that the specified writer is shutting down gracefully and no longer
+     * needs to be considered for calculating entries for the marks segment. This may not be called
+     * in the event that writer crashes.
+     *
+     * @param writerId The name of the writer. (User defined)
+     * @param stream The stream the writer was on.
+     */
+    void writerShutdown(String writerId, Stream stream);
+```
+In order to supply this Position the `EventStreamWriterImpl` would  internally track the offsets for the Segments it was writing to. To do this we would piggyback this information onto the acks it is already receiving from the server. This would involve making a backwards compatible change to the wire protocol to add a `currentSegmentWriteOffset` to `DataAppended`:
+
+```
+@Data
+public static final class DataAppended implements Reply, WireCommand {
+   final WireCommandType type = WireCommandType.DATA_APPENDED;
+   final UUID writerId;
+   final long eventNumber;
+   final long previousEventNumber;
+   final long currentSegmentWriteOffset;
+//...
+```
+
+A similar field would be required for `SegmentMerged`:
+```
+@Data
+public static final class SegmentsMerged implements Reply, WireCommand {
+   final WireCommandType type = WireCommandType.SEGMENTS_MERGED;
+   final long requestId;
+   final String target;
+   final String source;
+   final long newTargetWriteOffset;
+//...
+```
+
+This would expand the size of each of these messages by 8 bytes. It would be compatible because the absence of the field can be detected.
+
+## Server API Changes
+Besides returning the current offsets on merge and ack there are no changes required. In the event that a `SegmentMerged` is lost and the client reconnects to see the Segment is already merged, the client can synthesize a valid if conservative offset by calling `getSegmentLength()`.
+
+## Client API Changes
+Internally the writer needs to collect the offsets returned in the acks it is receiving from the server. When `noteTime` is invoked it sends the `timestamp` provided along with the latest offsets it has for the Segments it is writing to are sent to the Controller’s `noteTimestampFromWriter` method.
+
+For a transactional writer, it just passes the `timestamp` specified on commit to the Controller.
+
+Each Reader needs to read from the mark Segment and compare its position to the `streamCuts`. In the Reader Group state, a new Map needs to be added from Reader to `timestamp`. Every time a Reader advances beyond a `streamCut` (counting as its own any unassigned Segments) it updates its entry in the Reader Group state providing the `timestamp` it just passed. If a Reader loses its last Segment it removes itself from the Map.
```

Comment:
Here and in the surrounding paragraphs, it's talking about the implementation. But uses phrases like 'needs to' which we use because we are the ones implementing, but make no sense to a user because it sounds like they have to do this.

Reply:
@tkaitchuck removed